File: packages/typescript/src/core/adapters/local-secure.ts
/**
 * Local signer backed by an encrypted secret file.
 */

import type { LocalSecureWalletParams } from '../config.js'
import { LocalSigner } from './local.js'

export type SecretLoaderFn = (configDir: string, password: string, secretRef: string) => Uint8Array

export class LocalSecureSigner extends LocalSigner {
  constructor(
    params: LocalSecureWalletParams,
    configDir: string,
    password: string | undefined,
    network: string,
    secretLoader: SecretLoaderFn | undefined,
  ) {
    if (!password) throw new Error('Password required for local_secure wallets')
    if (!secretLoader) throw new Error('local_secure wallets require a configured secret loader')
    const privateKey = secretLoader(configDir, password, params.secret_ref)
    super(privateKey, network)
  }
}

File: packages/python/src/agent_wallet/core/adapters/local_secure.py
"""Local signer backed by an encrypted secret file."""

from __future__ import annotations

from collections.abc import Callable
from pathlib import Path

from agent_wallet.core.adapters.local import LocalSigner
from agent_wallet.core.config import LocalSecureWalletParams


class LocalSecureSigner(LocalSigner):
    """Local signer backed by an encrypted secret file."""

    def __init__(
        self,
        params: LocalSecureWalletParams,
        config_dir: str | Path,
        password: str | None,
        network: str | None,
        secret_loader: Callable[[str | Path, str, str], bytes] | None,
    ) -> None:
        if not password:
            raise ValueError("Password required for local_secure wallets")
        if secret_loader is None:
            raise ValueError("local_secure wallets require a configured secret loader")
        private_key = secret_loader(config_dir, password, params.secret_ref)
        super().__init__(private_key=private_key, network=network)
